server/src/main/java/org/apache/iotdb/db/engine/storagegroup/StorageGroupProcessor.java

```diff
@@ -1886,13 +1889,9 @@ private void loadUpgradedResources(List<TsFileResource> resources, boolean isseq
     resources.clear();
   }
 
-  public void merge(boolean fullMerge) {
-    writeLock();
-    try {
-      this.tsFileManagement.merge(fullMerge, tsFileManagement.getTsFileList(true),
-          tsFileManagement.getTsFileList(false), dataTTL);
-    } finally {
-      writeUnlock();
+  public void merge(boolean isFullMerge) {
+    for (long timePartitionId : timePartitionIdVersionControllerMap.keySet()) {
+      executeCompaction(timePartitionId, isFullMerge);
     }
```

Comment:
Now that the write lock is removed, is it possible that if more than one client issues concurrent merge requests, the result of ` tsFileManagement.forkCurrentFileList` and `tsFileManagement.setForceFullMerge` will be messed up, when the merge requests contain different arguments?

Reply:
The param `compactionMergeWorking` in `StorageGroupProcessor` ensures that there is only one compaction thread.

Reply:
I solve the problem in this way because I think it is also necessary to update the function of the command `merge`.

Reply:
Unfortunately, a simple if-else is barely enough for concurrency control, because the condition check and the flag set action are not atomic,  multiple threads may still enter that block concurrently.

Reply:
okay, I will add lock in the function executeCompaction().